packages/alfa-dom/src/matches.ts

```diff
@@ -178,15 +200,62 @@ function matchesClass(element: Element, selector: ClassSelector): boolean {
 /**
  * @see https://www.w3.org/TR/selectors/#type-selectors
  */
-function matchesType(element: Element, selector: TypeSelector): boolean {
+function matchesType(
+  element: Element,
+  context: Node,
+  selector: TypeSelector,
+  options: MatchesOptions
+): boolean {
   // https://www.w3.org/TR/selectors/#the-universal-selector
   if (selector.name === "*") {
     return true;
   }
 
+  if (matchesNamespace(element, context, selector, options) === false) {
+    return false;
+  }
+
   return element.localName === selector.name;
 }
 
+/**
+ * @see https://www.w3.org/TR/selectors/#type-nmsp
+ */
+function matchesNamespace(
+  element: Element,
+  context: Node,
+  selector: TypeSelector,
+  options: MatchesOptions
+): boolean {
+  if (selector.namespace === "*" || options.namespaces === undefined) {
+    return true;
+  }
+
+  const elementNamespace = getElementNamespace(element, context);
+
+  // As all elements in HTML5 will use the XHTML namespace,
+  // unless another namespace is explicitly specified, the
+  // only way to currently test the no-namespace selector
+  // (i.e. "|div") is to provide a disconnected context.
+  // https://www.w3.org/TR/selectors/#type-nmsp
+  if (selector.namespace === "" && elementNamespace === null) {
+    return true;
+  }
+
+  // The namespace must have been declared.
+  const declaration = options.namespaces.get(selector.namespace);
+
+  if (declaration === undefined) {
+    return false;
+  }
+
+  if (elementNamespace !== declaration) {
+    return false;
+  }
+
+  return true;
```

Comment:
Can be reduced to:

```ts
return elementNamespace === declaration;
```

I do like the explicitness of pulling out the declared namespace and checking if it's `undefined`. Maybe rename `declaration` to `declaredNamespace` though?